manim/scene/scene.py

```diff
@@ -849,8 +883,36 @@ def get_time_progression(
         if self.renderer.skip_animations and not override_skip_animations:
             times = [run_time]
         else:
-            step = 1 / config["frame_rate"]
-            times = np.arange(0, run_time, step)
+            if speedinfo is not None:
+                nodes = list(speedinfo.keys())
+                speeds = list(speedinfo.values())
+                if nodes[0] != 0:
+                    nodes.insert(0, 0)
+                if nodes[-1] != 1:
+                    nodes.append(1)
+                    speeds.append(speeds[-1])
+                times = np.array([])
+                on_node = 0
+
+                for dur, s in zip(np.diff(nodes), speeds):
```

Comment:
The code in this loop is rather hard to read. Could you please add some comments and more explicit variable names? (`dur` -> `duration`, `s` -> `speed`, etc.; `dt` is fine.)